cocos/renderer/backend/ProgramState.h

```diff
@@ -66,13 +69,27 @@ class ProgramState : public Ref
     inline std::vector<UniformBuffer>& getVertexUniformInfos() { return _vertexUniformInfos; }
     inline const std::vector<UniformBuffer>& getFragmentUniformInfos() const { return _fragmentUniformInfos; }
 
+    void setCallbackUniform(const backend::UniformLocation&, const UniformCallback &);
+
     //set textures
     void setTexture(const backend::UniformLocation& uniformLocation, uint32_t slot, backend::Texture* texture);
     void setTextureArray(const backend::UniformLocation& uniformLocation, const std::vector<uint32_t>& slots, const std::vector<backend::Texture*> textures);
 
     inline const std::unordered_map<int, TextureInfo>& getVertexTextureInfos() const { return _vertexTextureInfos; }
     inline const std::unordered_map<int, TextureInfo>& getFragmentTextureInfos() const { return _fragmentTextureInfos; }
-    
+    inline const std::unordered_map<UniformLocation, UniformCallback, UniformLocation>& getCallbackUniforms() const { return _callbackUniforms; }
+
+    class CC_DLL AutoBindingResolver {
+    public:
+        virtual ~AutoBindingResolver();
+        //TODO doc
+        virtual bool resolveAutoBinding(ProgramState *, const std::string &uniformName, const std::string &autoBinding) = 0;
+    protected:
+        AutoBindingResolver();
+    };
+
+    void setParameterAutoBinding(const std::string &, const std::string &);
```

Comment:
What's auto binding?

Reply:
Uniform fields in material file
https://github.com/cocos2d/cocos2d-x/blob/f0f732cc90c0ac18cbe2e4644f69c2bbfcc6a2bd/tests/cpp-tests/Resources/Materials/auto_binding_test.material#L16

which value is set on runtime, rather than set in the material file.